pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go

```diff
@@ -0,0 +1,561 @@
+	serviceName   = "kube-dns-upstream"
+	configDataKey = "Corefile"
+)
+
+// Interface contains functions for a NodeLocalDNS deployer.
+type Interface interface {
+	component.DeployWaiter
+	component.MonitoringComponent
+}
+
+// Values is a set of configuration values for the nodelocaldns component.
+type Values struct {
+	// Image is the container image used for NodeLocalDNS.
+	Image string
+	// VPAEnabled marks whether VerticalPodAutoscaler is enabled for the shoot.
+	VPAEnabled bool
+	// ForceTcpToClusterDNS enforces upgrade to tcp connections for communication between node local and cluster dns.
+	ForceTcpToClusterDNS bool
+	// ForceTcpToUpstreamDNS enforces upgrade to tcp connections for communication between node local and upstream dns.
+	ForceTcpToUpstreamDNS bool
+	// ClusterDNS
+	ClusterDNS string
+	// DNSServer
+	DNSServer string
+}
+
+// New creates a new instance of DeployWaiter for nodelocaldns.
+func New(
+	client client.Client,
+	namespace string,
+	values Values,
+) Interface {
+	return &nodeLocalDNS{
+		client:    client,
+		namespace: namespace,
+		values:    values,
+	}
+}
+
+type nodeLocalDNS struct {
+	client    client.Client
+	namespace string
+	values    Values
+}
+
+func (c *nodeLocalDNS) Deploy(ctx context.Context) error {
+	data, err := c.computeResourcesData()
+	if err != nil {
+		return err
+	}
+	return managedresources.CreateForShoot(ctx, c.client, c.namespace, ManagedResourceName, false, data)
+}
+
+func (c *nodeLocalDNS) Destroy(ctx context.Context) error {
+	return managedresources.DeleteForShoot(ctx, c.client, c.namespace, ManagedResourceName)
+}
+
+// TimeoutWaitForManagedResource is the timeout used while waiting for the ManagedResources to become healthy
+// or deleted.
+var TimeoutWaitForManagedResource = 2 * time.Minute
+
+func (c *nodeLocalDNS) Wait(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	return managedresources.WaitUntilHealthy(timeoutCtx, c.client, c.namespace, ManagedResourceName)
+}
+
+func (c *nodeLocalDNS) WaitCleanup(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	return managedresources.WaitUntilDeleted(timeoutCtx, c.client, c.namespace, ManagedResourceName)
+}
+
+func (c *nodeLocalDNS) computeResourcesData() (map[string][]byte, error) {
+	var (
+		hostPathFileOrCreate = corev1.HostPathFileOrCreate
+		registry             = managedresources.NewRegistry(kubernetes.ShootScheme, kubernetes.ShootCodec, kubernetes.ShootSerializer)
+
+		serviceAccount = &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "node-local-dns",
+				Namespace: metav1.NamespaceSystem,
+			},
+		}
+
+		podSecurityPolicy = &policyv1beta1.PodSecurityPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "gardener.kube-system.node-local-dns",
+				Labels: map[string]string{
+					"app": "node-local-dns",
```

Comment:
```suggestion
					v1beta1constants.LabelApp: LabelValue,
```

Reply:
Suggestion : You can use a function for all the cases which have the same labels.
```
func getLabels() map[string]string {
	return map[string]string{v1beta1constants.LabelApp: LabelValue}
}
```

Reply:
It's needed only on 3 components rest of them have different labels or have no label. So I suggest no need to add function.